stm32-modules/include/common/core/ads1115.hpp

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file ads1115.hpp
+ * @brief
+ * Driver file for the ADS1115 ADC. See ads1115.cpp for more details.
+ */
+
+#pragma once
+
+#include <concepts>
+#include <cstdint>
+#include <optional>
+#include <variant>
+
+namespace ADS1115 {
+
+/**
+ * @brief The way that the ADS1115 policy is structured assumes that each
+ * ADC instance is provided an instance of a Policy type that is aware of
+ * which ADC it is talking to, so there is no need to specify the address
+ * or any other enumeration of the ADC
+ */
+template <typename Policy>
+concept ADS1115Policy = requires(Policy& p, uint8_t u8, uint16_t u16,
+                                 uint32_t u32) {
+    // A function to mark that an ADS1115 was initialized.
+    { p.ads1115_mark_initialized() } -> std::same_as<void>;
+    // A function to check that an ADS1115 was initialized
+    { p.ads1115_check_initialized() } -> std::same_as<bool>;
+    // Acquire the mutex for this ADC. The mutex must be initialized with the
+    // policy, so it is always valid
+    { p.ads1115_get_lock() } -> std::same_as<void>;
+    // Release the mutex for this ADC. The mutex must be initialized with the
+    // policy, so it is always valid
+    { p.ads1115_release_lock() } -> std::same_as<void>;
+    // Arm this ADC for a read operation
+    { p.ads1115_arm_for_read() } -> std::same_as<bool>;
+    // Write a 16 bit register
+    { p.ads1115_i2c_write_16(u8, u16) } -> std::same_as<bool>;
+    // Read a 16 bit register
+    { p.ads1115_i2c_read_16(u8) } -> std::same_as<std::optional<uint16_t>>;
+    // Waits for a pulse from the ADC that was armed by this task. Maximum
+    // wait time is passed as a parameter.
+    { p.ads1115_wait_for_pulse(u32) } -> std::same_as<bool>;
+};
+
+enum class Error {
+    ADCTimeout = 1, /**< Timed out waiting for ADC.*/
+    I2CTimeout = 2, /**< Timed out waiting for I2C.*/
+    DoubleArm = 3,  /**< ADC already armed.*/
+    ADCPin = 4,     /**< Pin is not allowed.*/
+    ADCInit = 5     /**< ADC is not initialized.*/
+};
+
+template <ADS1115Policy Policy>
+class ADC {
+  public:
+    using ReadVal = std::variant<uint16_t, Error>;
+
+    ADC() = delete;
+    /**
+     * @brief Construct a new ADS1115 ADC
+     * @param[in] addr The I2C address of the ADC
+     * @param[in] id The ID of the ADC, which will link to one of the
+     * interrupts defined in \ref thermal_hardware.h
+     */
+    ADC(Policy& policy) : _policy(policy) {}
```

Comment:
not really critical but single-argument ctors should be marked `explicit` so they don't participate in implicit conversions